### PR TITLE
build: Set triton environment variables

### LINKED ISSUE #367: FMS-HF-Tuning QLoRA fine-tuning crashes because it cannot access `/.triton` in OpenShift
## Describe the bug

When running QLoRA on OpenShift, FMS-HF-Tuning crashes because it cannot access `/.triton`

## Platform

OpenShift AI, `quay.io/modh/fms-hf-tuning:v2.0.1`

## Sample Code

Running this configuration:
```
+ cat /mnt/config/config.json
{
  "training_data_path": "/mnt/output/dataset.json",
  "model_name_or_path": "/mnt/storage/model/mistral-7b-v0.3-gptq",
  "response_template": " ### Label:",
  "output_dir": "/mnt/output/fine-tuning",
  "save_model_dir": "/mnt/output/save-model-dir",
  "accelerate_launch_args": {
    "num_processes": 8,
    "num_machines": 1,
    "mixed_precision": "no",
    "dynamo_backend": "no",
    "downcast_bf16": "no",
    "main_training_function": "main",
    "rdzv_backend": "static",
    "same_network": true,
    "tpu_use_sudo": false,
    "use_fsdp": true
  },
  "num_train_epochs": 1,
  "per_device_train_batch_size": 4,
  "per_device_eval_batch_size": 4,
  "gradient_accumulation_steps": 4,
  "eval_strategy": "no",
  "save_strategy": "no",
  "learning_rate": 0.00001,
  "weight_decay": 0,
  "lr_scheduler_type": "cosine",
  "max_seq_length": 1024,
  "include_tokens_per_second": true,
  "dataset_text_field": "output",
  "use_flash_attn": true,
  "auto_gptq": [
    "triton_v2"
  ],
  "fp16": true,
  "gradient_checkpointing": true,
  "lora_alpha": 16,
  "max_steps": -1,
  "packing": false,
  "peft_method": "lora",
  "r": 4,
  "target_modules": [
    "all-linear"
  ],
  "torch_dtype": "float16",
  "warmup_ratio": 0.03
}
```

## Expected behavior

The container image should run fine-tuning without having to specify any extra environment variable

## Observed behavior

```
  File "/home/tuning/.local/lib/python3.11/site-packages/triton/runtime/cache.py", line 64, in __init__
    os.makedirs(self.cache_dir, exist_ok=True)
  File "<frozen os>", line 215, in makedirs
  File "<frozen os>", line 215, in makedirs
  File "<frozen os>", line 225, in makedirs
PermissionError: [Errno 13] Permission denied: '/.triton'
```

[pod.log](https://github.com/user-attachments/files/17290445/pod.41.log)

## Additional context

Work around is to define these environment variables, and make them point to a writable directory: 
```
            env:
            - name: TRITON_HOME
              value: "/mnt/output"
            - name: TRITON_DUMP_DIR
              value: "/mnt/output"
            - name: TRITON_CACHE_DIR
              value: "/mnt/output"
            - name: TRITON_OVERRIDE_DIR
              value: "/mnt/output"
```

### PR DESCRIPTION
**Pull request by willmj**
### Description of the change
Add triton environment variables for QLoRA tuning if random user in cluster.

### Related issue number

closes #367 

### How to verify the PR
Use image `docker-na-public.artifactory.swg-devops.com/wcp-ai-foundation-team-docker-virtual/sft-trainer:triton_env_vars_ubi9_py311.triton_env_vars` and run tuning as a random user without setting these variable manually.

### Was the PR tested

- [ ] I have added >=1 unit test(s) for every new method I have added.
- [x] I have ensured all unit tests pass

### CODE REVIEW

File: build/Dockerfile
Comment (anhuong) [inline build/Dockerfile:176]: Can you describe a little more what gets written to these paths and could this cause another ephemeral storage issue?

Comment (willmj) [reply build/Dockerfile:176]: Definitely! After training `llama2-70b-gptq`, the following files were written into /tmp/
```
$ $ du -bsh *
31G	output
5.2K	tmp5svfusys
5.2K	tmpivnfgqr_
86	torchelastic_f3_xoj2g
1.3M	triton_cache_dir
15K	twitter-complaints.json
```

Inside `/triton_cache_dir` are the following files:
```
$ du -bsh *
47K	00b59251c9395aa85e84d9109d00b918f52971b9654e0617a02cfa17a63def95
47K	0f72e49f3d6f51cbc1fad453cdd838886e55bc3676659e27b2b0e5509bcd567b
58K	11f223788a89c81fc65244b2d2b6ce5f330747301ea03f74af51dc40eeef64b6
47K	152ecf14b4c0c916b269be9ed5dfecca4bbf342e83abd7f2cffa5238191f14ee
50K	20cede54259f483813ff56dc0861d0601f66192a0336d9ff169885b60c8d399c
50K	2592cd0ef2049a0e24aa77d04da6652edb52a81c26524bfe6cb46c29f881d1b4
50K	2871408de0129abbba624b09054ac5d56b116455502ff5e846979e2f4ca21855
51K	29b662d19c7774b55a9d023d02d9dd0e4ef647f1ee7d18b8fbbb0a792d441755
51K	2e0c262613a6a77fd44968b74e06432ab612fb949ebfb305056e3e1037d2ece4
47K	3e2477233858d58dda1742f79d2c09bd5277e2c8611f188d8197223bccc72359
23K	41ce1f58e0a8aa9865e66b90d58b3307bb64c5a006830e49543444faf56202fc
74K	52ab7f05c665150fc61e5568c226637ae7a4e833ac1ef5706a9e5315f0401dc2
47K	52d3d47b343c81241fbbacd01fe17c84e96bd2424789e15c58e1237d34bc1eea
58K	5cf019e9bf065570457ec194793e2ee671b5f63b59800d2761e41084041b4c85
47K	694d7a4f90e5f4131d23dfeb244e3935b5d621bd48796151116c0c7e1d1a21e8
58K	70ac819b6ea5ab3fec34b0c3bb68da5860a6504b594403e662c055756ddd4b44
47K	79a25264b3198fd2099e1a634b8c90bc27a13d1cb7fe919be67f7fe667f46058
47K	80b65dbb7890a2410196d7f357311d91a687703811a496ca9431f7a0461136b0
58K	b8b8cfeb93af1f30698e341771dff8d295e3935754b30ec662e855a01281dc3a
72K	caf06e9c6a31f0577b4a91b4a70e55f768a9248c695bb5e80a9dbf592def76ac
47K	d051671f604ac9a166a6231573c2e5fa47adfca3a26b30e5a6e549846b996e0f
57K	d83bf161f11a5b53f3a512fba0e15dff3de0d29e42f705529b2732bf217da25f
18K	dcc301ce1ecd3cddf98f02d5108f2da91bec119dba08c319dd1b48a7e601e8bf
72K	e2935437a6d341867fffa29d33f1110f6318a3247e2f6826659a72858f26c049
57K	f430cb6af2431650ecdc7ecea410e167817302e6b14f483d26a8a6481efdec80
50K	fcacf88674fae7b35befe22f8c2fd6a3d6fa5204ab5a8fe85b1055e97a484917
```

These files are used during training to manage the cache, and I believe this is the [code](https://github.com/triton-lang/triton/blob/main/python/triton/runtime/cache.py) responsible.

Overall I don't think we'd run into any concerns with ephemeral storage as the files created by Triton are only 1.3MB+10.4kB. If there's a better place to write them we could do that as well.

**Review by anhuong — COMMENTED** [review]
LGTM, waiting to see if get any additional feedback on implications of setting env vars. This is only needed for QLoRA since using triton kernel but shouldn't hurt when set in other use cases. Ephemeral storage shouldn't be an issue as noted. 

Some other ideas we talked about on how to solve this problem:
- create /.triton with full open write access
- user can still override these values with their own

**Review by anhuong — APPROVED** [review]
Verifying one last thing, @willmj do these directories need to already be created? I don't think so and triton creates them but wanted to make sure

**Comment by willmj** [comment]
@anhuong Triton creates them, but it seems like most don't end up having anything written to them anyways.

**Pull request merged**
